### PR TITLE
statusandheaders: if setting a header fails on Headers object, retry with encoded header

### LINKED ISSUE #81: Headers with non-ASCII characters silently disappear
# Steps to reproduce
Take [this example WARC](https://github.com/user-attachments/files/17348821/example.warc.gz), produced with `wget --no-verbose -O /dev/null --max-redirect=0 --warc-file=example 'https://wiki.archlinux.org/title/AUR_Metadata_(%E6%97%A5%E6%9C%AC%E8%AA%9E)'`. It contains the line:

```
Location: https://wiki.archlinux.jp/index.php/Arch_User_Repository?rdfrom=https%3A%2F%2Fwiki.archlinux.org%2Findex.php%3Ftitle%3DAUR_Metadata_%28%25E6%2597%25A5%25E6%259C%25AC%25E8%25AA%259E%29%26redirect%3Dno#AUR_メタデータ
```

## Actual behavior
However, trying to read that header produces `null`:

```js
import {WARCParser} from 'warcio'

for await (const record of new WARCParser(process.stdin)) {
	if (record.warcType !== 'response') continue
	console.log(record.httpHeaders.headers.get('Location'))
}
// => null
```

## Expected behavior

I am unclear on what the “correct” behavior here is ([RFC 7230](https://www.rfc-editor.org/rfc/rfc7230#section-3.2.4) leads me to believe that the server probably shouldn't have sent a non-ASCII value to begin with, but since it did this header should have been interpreted as ISO-8859-1 mojibake; I can’t find anything in the Fetch spec that describes these values as anything other than a “byte sequence” though in practice Chrome, Firefox, and Node all limit values to U+00FF and below). Unfortunately since this behavior is in the wild I have to handle it anyway.

However, I *do* know that the header silently disappearing is very confusing and definitely not what I want to happen; it makes this problem hard to debug. I would have much preferred a crash, which would have let me immediately identify the cause.

# Workaround

Setting `keepHeadersCase: true` when creating the `WARCParser` uses a plain [`Map`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) instead of a [`Headers`](https://developer.mozilla.org/en-US/docs/Web/API/Headers), which doesn’t have the same limitation on values. Unfortunately this loses the case-insensitive aspect.

# Cause
The silent disappearance appears to be because of this code:

https://github.com/webrecorder/warcio.js/blob/fb1ff9c3d88a54fdc0acb2670ca25966b04fa95e/src/lib/statusandheaders.ts#L142-L150

...which is swallowing the error (“TypeError: Cannot convert argument to a ByteString because the character at index 204 has a value of 12513 which is greater than 255.”), along with anything else that might go wrong.

This behavior seems to have been there since 40610030ca54b7a0f0713aad52fc18ef60578721:

https://github.com/webrecorder/warcio.js/blob/40610030ca54b7a0f0713aad52fc18ef60578721/src/statusandheaders.js#L80-L85

...and since that was the initial commit, there's no obvious explanation for why it exists.

# Suggested fix
This is enough for my use case:
```diff
-try { 
+// If this crashes with an error like "String contains non ISO-8859-1 code point"
+// or "character has a value which is greater than 255", the response contains a
+// header which cannot be represented by the Headers class; enable `keepHeadersCase`
+// to use a Map (which does not have this restriction) instead.
   if (canAppend && name.toLowerCase() === "set-cookie") { 
     headers.append(name, value); 
   } else { 
     headers.set(name, value); 
   } 
-} catch (_e) { 
-  // ignore 
-} 
```
However, this may be a problem for users who have code that depends on the current behavior, are fine with that, and don't want to have to rewrite their code to handle lookups in a Map. For that use case it might be necessary to thread through a flag (call it `discardInvalidHeaders`, say) to keep the current behavior. (Either way I think the defaults should change, which would be a bump to the semver major version.)

### PR DESCRIPTION
**Pull request by ikreymer**
instead of skipping invalid headers:
- encode header value as latin-1 for use with Headers object
- encode back to original when serializing
fixes #81

### CODE REVIEW

**Comment by ikreymer** [comment]
Trying to decide if we want to go with encodeURI or with latin-1 encoding here... I think latin-1 is technically more 'correct', though encodeURI() more portable / easier to understand. As mentioned in #81, fetch() does the equivalent of latin-1 / iso-8859-1, so maybe we do go with that?

**Comment by tw4l** [comment]
> Trying to decide if we want to go with encodeURI or with latin-1 encoding here... I think latin-1 is technically more 'correct', though encodeURI() more portable / easier to understand. As mentioned in #81, fetch() does the equivalent of latin-1 / iso-8859-1, so maybe we do go with that?

I can see the arguments either way but my vote would be for ISO-8859-1, since that's what the HTTP spec seems to have originally supported and is consistent with fetch.

**Comment by ikreymer** [comment]
> > Trying to decide if we want to go with encodeURI or with latin-1 encoding here... I think latin-1 is technically more 'correct', though encodeURI() more portable / easier to understand. As mentioned in #81, fetch() does the equivalent of latin-1 / iso-8859-1, so maybe we do go with that?
> 
> I can see the arguments either way but my vote would be for ISO-8859-1, since that's what the HTTP spec seems to have originally supported and is consistent with fetch.

Updated it to the latin-1 encoding again.

File: src/lib/statusandheaders.ts [hunk 1]
```diff
@@ -202,6 +205,16 @@ function splitRemainder(str: string, sep: string, limit: number) {
   return newParts;
 }
 
+// ===========================================================================
+function encodeLatin1(value: string) {
+  const buf = new TextEncoder().encode(value);
+
+  let str = "";
+  buf.forEach((x) => (str += String.fromCharCode(x)));
+  //buf.forEach(x => str += x < 128 ? String.fromCharCode(x) : `\\x${x}`);
```
Comment (tw4l) [inline src/lib/statusandheaders.ts]: I wonder if using [buffer.transcode()](https://nodejs.org/api/buffer.html#buffer_buffer_transcode_source_fromenc_toenc) might be a good idea here, as there might be edge cases where `TextEncoder().encode(value)` might encode certain characters as two bytes?

**Review by tw4l — APPROVED** [review]
One suggestion to take or leave re: encoding, otherwise looks good and thanks for the tests

File: src/lib/statusandheaders.ts (hunk 1 above)
Comment (ikreymer) [reply src/lib/statusandheaders.ts]: Hmm, yeah, though buffer is not available in the browser, so would have to use the equivalent.

Comment (tw4l) [reply src/lib/statusandheaders.ts]: Ah true, so used to working with node I forgot! It is also an edge case within an edge case, so I trust how much attention you want to give this

Comment (ikreymer) [reply src/lib/statusandheaders.ts]: I guess if we change this, it will break this assertion from the test case:

```
function decodeLatin1(buf) {
  let str = "";
  for (let i = 0; i < buf.length; i++) {
    str += String.fromCharCode(buf[i]);
  }
  return str;
}
d = decodeLatin1(new TextEncoder().encode("https://wiki.archlinux.jp/index.php/Arch_User_Repository?rdfrom=https%3A%2F%2Fwiki.archlinux.org%2Findex.php%3Ftitle%3DAUR_Metadata_%28%25E6%2597%25A5%25E6%259C%25AC%25E8%25AA%259E%29%26redirect%3Dno#AUR_メタデータ"));

// this matches the headers from node fetch:
a = await fetch("https://wiki.archlinux.org/title/AUR_Metadata_(%E6%97%A5%E6%9C%AC%E8%AA%9E)", {redirect: "manual"})
assert(a.headers.get("location") === d)
```

If we do: `decodeLatin1(Buffer.from("https://wiki.archlinux.jp/index.php/Arch_User_Repository?rdfrom=https%3A%2F%2Fwiki.archlinux.org%2Findex.php%3Ftitle%3DAUR_Metadata_%28%25E6%2597%25A5%25E6%259C%25AC%25E8%25AA%259E%29%26redirect%3Dno#AUR_メタデータ", "ascii"));` it will give a different result than what node fetch is doing..

Comment (ikreymer) [reply src/lib/statusandheaders.ts]: Resolved this in the most correct way, I think. Still encoding utf8-latin1 to store in Headers object, but then reencoding back when serializing

**Review by tw4l — APPROVED** [review]
